BranchSDK/BNCReferringURLUtility.m

```diff
@@ -0,0 +1,205 @@
+    NSArray *validURLQueryParameters = @[BRANCH_REQUEST_KEY_REFERRER_GBRAID, BRANCH_REQUEST_KEY_GCLID];
+    if ([validURLQueryParameters containsObject:param]) {
+        return YES;
+    } else {
+        return NO;
+    }
+}
+
+- (BNCUrlQueryParameter *)findUrlQueryParam:(NSString *)paramName {
+    if ([self.urlQueryParameters.allKeys containsObject:paramName]) {
+        return self.urlQueryParameters[paramName];
+    } else {
+        BNCUrlQueryParameter *param = [BNCUrlQueryParameter new];
+        param.name = paramName;
+        return param;
+    }
+}
+
+- (NSTimeInterval)defaultValidityWindowForParam:(NSString *)paramName {
+    if ([paramName isEqualToString:BRANCH_REQUEST_KEY_REFERRER_GBRAID]) {
+        return 2592000; // 30 days = 2,592,000 seconds
+    } else {
+        return 0; //Default, means indefinite.
+    }
+}
+
+- (NSMutableDictionary *)serializeToJson:(NSMutableDictionary<NSString *, BNCUrlQueryParameter *> *)urlQueryParameters {
+    NSMutableDictionary *json = [NSMutableDictionary new];
+    
+    for (BNCUrlQueryParameter *param in urlQueryParameters.allValues) {
+        NSMutableDictionary *paramDict = [NSMutableDictionary new];
+        paramDict[BRANCH_URL_QUERY_PARAMETERS_NAME_KEY] = param.name;
+        paramDict[BRANCH_URL_QUERY_PARAMETERS_VALUE_KEY] = param.value;
+        paramDict[BRANCH_URL_QUERY_PARAMETERS_TIMESTAMP_KEY] = param.timestamp;
+        paramDict[BRANCH_URL_QUERY_PARAMETERS_IS_DEEPLINK_KEY] = @(param.isDeepLink);
+        paramDict[BRANCH_URL_QUERY_PARAMETERS_VALIDITY_WINDOW_KEY] = @(param.validityWindow);
+        
+        json[param.name] = paramDict;
+    }
+    
+    return json;
+}
+
+- (NSMutableDictionary<NSString *, BNCUrlQueryParameter *> *)deserializeFromJson:(NSDictionary *)json {
+    NSMutableDictionary *result = [NSMutableDictionary new];
+    
+    for (id temp in json.allValues) {
+        if ([temp isKindOfClass:NSDictionary.class] || [temp isKindOfClass:NSMutableDictionary.class]) {
+            NSDictionary *paramDict = (NSDictionary *)temp;
+            BNCUrlQueryParameter *param = [BNCUrlQueryParameter new];
+            param.name = paramDict[BRANCH_URL_QUERY_PARAMETERS_NAME_KEY];
+            param.value = paramDict[BRANCH_URL_QUERY_PARAMETERS_VALUE_KEY];
+            param.timestamp = paramDict[BRANCH_URL_QUERY_PARAMETERS_TIMESTAMP_KEY];
+            param.validityWindow = [paramDict[BRANCH_URL_QUERY_PARAMETERS_VALIDITY_WINDOW_KEY] doubleValue];
+
+            if (paramDict[BRANCH_URL_QUERY_PARAMETERS_IS_DEEPLINK_KEY] != nil) {
+                param.isDeepLink = ((NSNumber *)paramDict[BRANCH_URL_QUERY_PARAMETERS_IS_DEEPLINK_KEY]).boolValue;
+            } else {
+                param.isDeepLink = NO;
+            }
+            
+            result[param.name] = param;
+        }
+    }
+    
+    return result;
+}
+
+//Turns an old gbraid into a new BNCUrlQueryParameter and saves it for use, then clears it.
+-(void)checkForAndMigrateOldGbraid {
```

Comment:
Let's either clear the old value, or check for a more recent value first. Don't want to migrate more than once.